qutebrowser/misc/sql.py

```diff
@@ -29,23 +29,73 @@
 
 class SqlError(Exception):
 
-    """Raised on an error interacting with the SQL database."""
+    """Raised on an error interacting with the SQL database.
 
-    pass
+    Attributes:
+        environmental: Whether the error is likely caused by the environment and
+                       not a qutebrowser bug.
+    """
+
+    def __init__(self, msg, environmental=False):
+        super().__init__(msg)
+        self.environmental = environmental
```

Comment:
If you can think of a better name for this, let me know!

Reply:
Regardless of the name, catching exceptions based on a flag is rather unorthodox. In conventional Python, this would be something like
```python
class SqlError(Exception):
    pass

class SqlEnvironmentalError(SqlError):
    pass

...

try:
    do_something()
except SqlEnvironmentalError:
    handle_gracefully()
except SqlError:
    log_and_reraise()  # or don't catch SqlError at all
```

But since the _environmental_ errors are "normal" and the other indicate bugs in qutebrowser, the naming should probably emphasize the latter: e.g. `SqlProgrammingError` or `SqlAssertionError`. See also [sqlite3 exceptions](https://docs.python.org/3/library/sqlite3.html#exceptions).

Also, since `SqliteError` below is never environmental, it should inherit from the `SqlProgrammingError` or `SqlAssertionError`.

Reply:
I've thought about that, but then the code raising the exceptions would either need to check the error codes every time, or we'd need another `_raise_sqlite_error()` function or so raising the actual exception - neither seemed very appealing.

> Also, since `SqliteError` below is never environmental, it should inherit from the `SqlProgrammingError` or `SqlAssertionError`.

Huh? It is environmental depending on the sqlite error code.

Reply:
Ouch, I totally missed the last line of the `SqliteError`'s `__init__` method because Github shows a comment just above it...

Anyway, in the current code there is this recurring pattern:
```python
try:
    do_something()
except sql.SqlError as e:
￼    if e.environmental:
￼        handle_error()
    else:
￼        raise
```
To me it doesn't seem appealing either, much less than the hypothetical `_raise_sqlite_error()` function...

Reply:
We already do have a `_raise_sqlite_error` function, called [`from_query`](https://github.com/qutebrowser/qutebrowser/pull/3073/files#diff-b171c554543e2419a0e422c8ca89c7d7R82) :stuck_out_tongue_winking_eye: 

Point being, if we already need a factory function, maybe that could choose what type to throw (though it might not be appropriate as a class method anymore).